### ISSUE TITLE
Cannot build SQL expression NOT(x IS NULL) on Postgres

### ISSUE THREAD

**Issue by SebAlbert** [sqlalchemy/sqlalchemy#6612]
I am on Postgres 12 (via psycopg) on SQLAlchemy 1.4.15, but I don't think this is too relevant.

I want to create the expression `NOT (x IS NULL)`, but I always get `x IS NOT NULL`.
Example code (even without DB connection, so not Postgres-specific):

```
import sqlalchemy as sa
t = sa.table("tbl", sa.column("x"))
print(sa.select(t.c.x).where(sa.not_(t.c.x.is_(None))))
```

I explicitly need `NOT (x IS NULL)` with composite (i.e., tuple / table-valued) types.
The following might surprise you in Postgres:
```SELECT (NULL, FALSE) IS NULL, (NULL, FALSE) IS NOT NULL```
But this is documented behaviour:
https://www.postgresql.org/docs/current/functions-comparison.html
> If the expression is row-valued, then IS NULL is true when the row expression itself is null or when all the row's fields are null, while IS NOT NULL is true when the row expression itself is non-null and all the row's fields are non-null.

**Comment by CaselIT** [comment]
Hi,

I remembered asking a similar question. Here it is: https://github.com/sqlalchemy/sqlalchemy/issues/5256

tldr: you can use `.self_group()`

**Comment by SebAlbert** [comment]
Thanks, I did not find the existing issue via search, but maybe my search terms were inadequate.
Since your issue was closed to be tracked in, #3566, which in turn is marked resolved, should we keep one open as use case? Other than that, feel free to close.

**Comment by CaselIT** [comment]
@zzzeek not sure if we want to implement something specific for this. Otherwise we can move to discussion.

> 
> 
> Thanks, I did not find the existing issue via search, but maybe my search terms were inadequate.
> Since your issue was closed to be tracked in, #3566, which in turn is marked resolved, should we keep one open as use case? Other than that, feel free to close.

Don't worry, searching is sometimes hard. Regarding #3566 I don't know if we have actually implemented that case, or if it's slipped between all other stuff

**Comment by zzzeek** [comment]
in #3566 it wasn't that clear to me what was going on but in this case I'd want to see if row valued types can signal to the compiler and/or expression construction to group in this way.

Can I have a real-world example to work with ?

**Comment by CaselIT** [comment]
this is an example: https://github.com/sqlalchemy/sqlalchemy/issues/5256#issue-600623503
```sql
select *
from some_table
where some_table is not null
```
vs
```sql
select *
from some_table
where not (some_table is null)
```

the return values is different if a row has all null values vs some null values vs no null values

**Comment by zzzeek** [comment]
OK so do we need to just add an implicit self_group() to table_valued when used in an IS context?   we do something like this right now for array.any().


```python
>>> from sqlalchemy import table, column, select
>>> s = table("some_table", column("x"))
>>> stmt = select(s).where(~(s.table_valued() == None))
>>> print(stmt)
SELECT some_table.x 
FROM some_table 
WHERE some_table IS NOT NULL
>>> stmt = select(s).where(~(s.table_valued() == None).self_group())
>>> print(stmt)
SELECT some_table.x 
FROM some_table 
WHERE NOT (some_table IS NULL)
```

**Comment by zzzeek** [comment]
somehting like this can sort of do it, just parenthesizes every time, not sure if that's OK

```diff
diff --git a/lib/sqlalchemy/sql/sqltypes.py b/lib/sqlalchemy/sql/sqltypes.py
index 4f8654afd9..0dde2d30f5 100644
--- a/lib/sqlalchemy/sql/sqltypes.py
+++ b/lib/sqlalchemy/sql/sqltypes.py
@@ -3199,6 +3199,28 @@ class TableValueType(HasCacheKey, TypeEngine):
             for elem in elements
         ]
 
+    class Comparator(TypeEngine.Comparator):
+        def operate(self, op, *other, **kwargs):
+            if op in (operators.eq, operators.is_):
+                return elements.BinaryExpression(
+                    self.expr,
+                    coercions.expect(roles.ConstExprRole, other[0]),
+                    operators.is_,
+                    negate=None,
+                    type_=Boolean,
+                ).self_group()
+            elif op in (operators.ne, operators.is_not):
+                return elements.BinaryExpression(
+                    self.expr,
+                    coercions.expect(roles.ConstExprRole, other[0]),
+                    operators.is_not,
+                    negate=None,
+                    type_=Boolean,
+                )
+            else:
+                return super(TableValueType.Comparator, self).operate(op, *other, **kwargs)
+
+    comparator_factory = Comparator
 
 class MatchType(Boolean):
     """Refers to the return type of the MATCH operator.
```

**Comment by SebAlbert** [comment]
Does this also apply to `tuple_(...)` constructs, or could it simply be extended to them as well?

**Comment by zzzeek** [comment]
we'd have to include that too.

what we're doing here is gathering requirements.  so , Python examples using tuple_(), table_valued() etc that we can turn into test cases, would be helpful.

**Comment by SebAlbert** [comment]
Would this count for a test case for `Tuple`?

```
import sqlalchemy as sa
print(sa.select(sa.not_(sa.tuple_(sa.null(), sa.false()).is_(None))))
```
(you would need to have this expected to be rendered as `SELECT NOT((NULL, false) IS NULL AS anon_1` as opposed to `SELECT (NULL, false) IS NOT NULL AS anon_1` or similar)

**Comment by zzzeek** [comment]
yup, that can work
